docs/content/docs/concepts/Breaking Changes and Supported Patterns.md

```diff
@@ -0,0 +1,293 @@
+
+These same variance annotations and rules can be applied to all types in the package, defaulting to invariant if not marked.
+Classes can default to "out" when marked with `@sealed` (which [already exists](https://api-extractor.com/pages/tsdoc/tag_sealed/)).
+Classes which are not `@sealed` are invariant by default since almost any modification to them could break User Code.
+The normal variance composition rules apply, so an `out` interface can only take `in` or invariant types as parameters to any methods on it.
+
+`const` module members can be typed as "out" unless documented otherwise.
+
+> **_NOTE:_** Our current type tests attempt to validate all types as if they were invariant.
+> The "out" aspect is covered by the backwards compatibility testing and the "in" is covered by the forwards compatibility testing.
+> If we adopted this convention, annotating our types could cause the type tests to omit the corresponding check allowing for some tooling based assistance with these rules for the package side.
+> It is theoretically possible to provide tooling level assistance about this for the user code side as well.
+> There isn't a clear robust solution, but many common cases may be coverable with not too complex custom linter rule.
+
+> **_NOTE:_** This section focuses on type level details, however this same policy can help with clarifying what's supported even if it's not modeled by the type systems. For example an interface that is documented as requiring an integer can be broadened to accept floating point values if it's a member of an "in" interface, but not for an "out" one. This reasoning is useful, even if both cases just use the type `number`.
+
+> **_NOTE:_** Another way to get some tooling support for this is to transform the types when exporting them either manually in the source or as part of publishing. For example if API-extractor could replace `@out` interfaces with types that are not constructable (extend a class with a protected member and private constructor).
+
+## Optional Function Arguments
+
+TypeScript has soundness issues with functions and optional arguments.
+
+For example:
+
+```typescript
+// Example function with an optional argument
+function processItem(item: string, override?: string): string {
+  return override ?? `processed: ${item}`;
+}
+
+function processItems(items: string[], f: (item: string) => string): string[] {
+  // When an function which requires two arguments is needed (like with Array.map, since it passes the item and the index),
+  // a function that requires fewer can be used.
+  return items.map(f);
+}
+
+// Functions with optional arguments can be assigned to types missing those arguments, as is done here.
+// When combined with allowing passing more arguments than needed, optional arguments can be passed incorrectly typed and unexpected data.
+// In this case the array ends up [1, 2, 3]: it's not even an array of strings like its type requires.
```

Comment:
Nit: It might be worth explicitly calling out that `map` passes the array index as the second argument. It took me a bit to remember this was the case, and to understand what was going on here.

Reply:
Also, shouldn't this be `[0,1,2]`?